### PR TITLE
Ian UI d2 1418 cstg keypair data

### PR DESCRIPTION
**Pull request by Ian-Nara**
Currently I have just tested with unit tests and using postman to list, then add,update, relist, etc.

### CODE REVIEW

File: src/main/java/com/uid2/admin/Main.java
Comment (Ian-Nara) [inline src/main/java/com/uid2/admin/Main.java:179]: How should these parameters be resolved?

Comment (sunnywu) [reply src/main/java/com/uid2/admin/Main.java:179]: @thomasm-ttd - the reason why we need to know if it's UID2/EUID scope and Prod/integ Env is to add these info into the private/public key pair for CSTG as you asked me a few weeks ago and i documented this as part of requirement for this ticket [here](https://atlassian.thetradedesk.com/confluence/display/UID2/UID2-1374+Proposal%3A+Data+Schema+and+Endpoint+Changes+for+CSTG+Support). Could you help answer how we go about determining these params in admin service?

File: src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java
Comment (sunnywu) [inline src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java:655]: you may keep these codes but currently we are not going to enable CSTG for private operators for now. So we do not need to provide any data in getClientSideKeypairs for private oeprator for now. 

We either simply don't generate site-specific data for keypair data (always keep keypair data as GlobalScope not SiteScope), or always return empty data for getClientSideKeypairs method.

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (sunnywu) [inline src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java:3]: as mentioned in my other comment, we won't enable CSTG for private operator so we don't need to generate PrivateSiteData here for private operator for now.

File: src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java:655]: Ok I disabled this with a comment in PrivateSiteDataSyncJob.java
`//        keypairSyncJob.execute(); Keypair sync Job should not be executed until we want to enable CSTG for private operators`
if that works for you.

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java:3]: Ok I disabled this with a comment in PrivateSiteDataSyncJob.java
//        keypairSyncJob.execute(); Keypair sync Job should not be executed until we want to enable CSTG for private operators
if that works for you.

File: src/main/java/com/uid2/admin/Main.java
Comment (thomasm-ttd) [reply src/main/java/com/uid2/admin/Main.java:179]: There is a branch waiting to be merged (tjm-UID2-1275-key-formatting) that changes the config parameters to "admin_key_prefix": "UID2-A-L-"
The UID2 is the identity scope, and the -L- is the environment: L = Local, P = prod, I = Integ

File: src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java
Comment (sunnywu) [reply src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java:655]: looks good!

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (sunnywu) [reply src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java:3]: looks good, do remember we only allow cstg api call for public (not private) operator and only public operator should load keystore data (if and when you will work on operator to load keypair data)

**Comment by sunnywu** [comment]
please don't merge this to master until Thomas advises us what to do in Slack

File: src/main/java/com/uid2/admin/util/InputUtil.java
Comment (jon8787) [inline src/main/java/com/uid2/admin/util/InputUtil.java:3]: is this class copied? Can we move it to shared instead?

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:210]: can we call toJson() from here?

File: src/main/java/com/uid2/admin/store/writer/ClientSideKeypairStoreWriter.java
Comment (jon8787) [inline src/main/java/com/uid2/admin/store/writer/ClientSideKeypairStoreWriter.java:30]: is this the same as toJsonWithPrivateKey? can we re-use?

File: src/test/java/com/uid2/admin/vertx/test/ServiceTestBase.java
Comment (jon8787) [inline src/test/java/com/uid2/admin/vertx/test/ServiceTestBase.java:183]: do we have to mock keypairSnapshot? can we not rely on keypairProvider.getSnapshot() ?

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:88]: why do we need to normalize the email?

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:156]: - Does this log the exception?
- I wonder if all these try/catch/500s are necessary. The generic error handler may already be doing this.

File: src/main/java/com/uid2/admin/util/InputUtil.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/util/InputUtil.java:3]: Yep, but just one method copied from the larger "InputUtil" class in uid2-client-java. Moving normalization logic to shared is part of [UID2-1206](https://atlassian.thetradedesk.com/jira/browse/UID2-1206). Once that is moved, the only change needed in Admin should be to delete InputUtil and change the import statement to shared. Assuming this contact is for human consumption only I'd say normalization is not necessary, however could be useful for catching typos.

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:88]: I suppose it isn't necessary, but it can provide some amount of typo protection in the event the email can't be normalized. If normalization succeed then it may just look nicer.

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:156]: Looks like it can either be caught or thrown up to setupRoutes. If throw up to setupRoutes have to catch there since it's an overrride so can't change the signature to throw an exception. I think its best to keep it this way.

File: src/main/java/com/uid2/admin/store/writer/ClientSideKeypairStoreWriter.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/store/writer/ClientSideKeypairStoreWriter.java]: Yeah we can do that. I suppose the ClientSideKeypairStoreWriter would be the most appropriate place for the keypair serialization logic? Will move there as static methods.

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java]: Refactored to remove duplicated code but maintain consistent json orderering across endpoints.

File: src/test/java/com/uid2/admin/vertx/test/ServiceTestBase.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/admin/vertx/test/ServiceTestBase.java]: Changed mock snapshot to real snapshot

File: src/main/java/com/uid2/admin/Main.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/Main.java]: I have updated this to use the prefix strategy of tjm-UID2-1275-key-formatting. I think the two branches should stay seperate, though. This branch has no dependency on 1275 and remains only additions.

File: src/main/java/com/uid2/admin/util/InputUtil.java
Comment (jon8787) [reply src/main/java/com/uid2/admin/util/InputUtil.java:3]: 1206 might not get done (see comments there). If you look at uid2-shared commits, there have been a few changes (eg by @cYKatherine ) where we have moved code to shared so it can be used by >1 project. Since we don't want to move all of InputUtil, we could use a different class in shared and make InputUtil depend on that

Comment (jon8787) [reply src/main/java/com/uid2/admin/util/InputUtil.java:3]: Discussed in slack, let's remove normalization for now

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (jon8787) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:163]: Oh it's a checked exception? That's annoying. Fine for now, as long as these get logged

File: src/main/java/com/uid2/admin/Main.java
Comment (sunnywu) [reply src/main/java/com/uid2/admin/Main.java]: that's ok for now

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (sunnywu) [inline src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java:59]: it's a bit not nice that there's a prefix for every key type but the IdentityScope in the prefix should be set once and reused in all of the key type prefixes. I guess room for improvement in the future.

File: src/main/java/com/uid2/admin/secret/SecureKeypairGenerator.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/secret/SecureKeypairGenerator.java]: removed.

File: src/main/java/com/uid2/admin/secret/IKeypairGenerator.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/secret/IKeypairGenerator.java]: changed the name

File: src/main/java/com/uid2/admin/secret/SecureKeypairGenerator.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/secret/SecureKeypairGenerator.java]: removed

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java]: reformatted

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java]: deleted the unneeded try catch

File: src/main/java/com/uid2/admin/util/InputUtil.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/util/InputUtil.java]: removed

File: src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java]: logs explicit error message + the exception

File: src/main/java/com/uid2/admin/Main.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/Main.java:147]: Should we have it just be created manually, then?

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (sunnywu) [inline src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java:106]: i would suggest to comment out all the codes you have added as we aren't generating keypair data for private operator yet for now.

**Comment by sunnywu** [comment]
Couple of minor comments but look ok to me. WIll leave it for @mcollins-ttd to review and approve.

File: src/main/java/com/uid2/admin/job/jobsync/key/ClientSideKeypairSyncJob.java
Comment (mcollins-ttd) [inline src/main/java/com/uid2/admin/job/jobsync/key/ClientSideKeypairSyncJob.java:1]: Does this mean every private operator gets all the client-side keypairs? We don't want private operators to get any client-side keypairs.

File: src/test/java/com/uid2/admin/vertx/ClientSideKeypairServiceTest.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/admin/vertx/ClientSideKeypairServiceTest.java]: added some letters

File: src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java
Comment (Ian-Nara) [reply src/test/java/com/uid2/admin/util/PrivateSiteUtilTest.java]: added some letters

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java]: commented the rest of this stuff out in PrivateSiteDataSyncJob

File: src/main/java/com/uid2/admin/job/jobsync/key/ClientSideKeypairSyncJob.java
Comment (Ian-Nara) [reply src/main/java/com/uid2/admin/job/jobsync/key/ClientSideKeypairSyncJob.java:1]: The job has not been added/enabled in PrivateSiteDataSyncJob now so it will not happen

File: src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java
Comment (sunnywu) [inline src/main/java/com/uid2/admin/job/jobsync/PrivateSiteDataSyncJob.java:24]: might as well comment these import statements  out

**Comment by sunnywu** [comment]
Looks ok, please resolve the branch conflict first. And ask @cody-constine-ttd if we should merge these to master at this point after this branch is ready to merge.

File: webroot/js/main.js
Comment (mcollins-ttd) [inline webroot/js/main.js:40]: Is this intentional?

**Comment by mcollins-ttd** [comment]
Looks like this needs to pull in the latest uid2-shared?

Sorry to be a pain, but would we be able to remove the commented-out/unused code? The code will still be in the commit history if we need it in the future, so can be easily retrieved.

I think this is everything:

- PrivateSiteDataSyncJob.java - commented-out code
- ClientSideKeypairSyncJob - unused
- ClientSideKeypairStoreFactory - unused
- PrivateSiteUtil.getClientSideKeypairs - unused
- Any tests for the above

File: webroot/js/main.js
Comment (Ian-Nara) [reply webroot/js/main.js]: nope

**Review by mcollins-ttd — APPROVED** [review]

**Pull request merged**
